# VLM Testing Notebook
A quick series of tests to determine the performance of Visual-language Models (VLM).

These tests were meant to be quick and subjective. If data-driven comparisons are needed, see one of the following sites:
- [Vision Arena](https://lmarena.ai/leaderboard/vision/ocr)
- [OpenVLM Leaderboard](https://huggingface.co/spaces/opencompass/open_vlm_leaderboard)

Two documents were used for these tests:
- A receipt with a number of line items, subtotals, taxes, and a final total. Find the original here: [Receipt](https://cas-bridge.xethub.hf.co/xet-bridge-us/6641997de7d4af2dcc8c77ce/1300fa843d9eb4d3e1e554a8fae6ff59c5d3eeb64c5fe6b9fefa458ddaf4bf39?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Content-Sha256=UNSIGNED-PAYLOAD&X-Amz-Credential=cas%2F20251104%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20251104T163433Z&X-Amz-Expires=3600&X-Amz-Signature=4f56fdd310693e88abb2865dac1e029a56a4652057efd7ce321d593b821c8afa&X-Amz-SignedHeaders=host&X-Xet-Cas-Uid=6750ae8de1d1bd01d01b13c7&response-content-disposition=inline%3B+filename*%3DUTF-8%27%27receipt_00008.png%3B+filename%3D%22receipt_00008.png%22%3B&response-content-type=image%2Fpng&x-id=GetObject&Expires=1762277673&Policy=eyJTdGF0ZW1lbnQiOlt7IkNvbmRpdGlvbiI6eyJEYXRlTGVzc1RoYW4iOnsiQVdTOkVwb2NoVGltZSI6MTc2MjI3NzY3M319LCJSZXNvdXJjZSI6Imh0dHBzOi8vY2FzLWJyaWRnZS54ZXRodWIuaGYuY28veGV0LWJyaWRnZS11cy82NjQxOTk3ZGU3ZDRhZjJkY2M4Yzc3Y2UvMTMwMGZhODQzZDllYjRkM2UxZTU1NGE4ZmFlNmZmNTljNWQzZWViNjRjNWZlNmI5ZmVmYTQ1OGRkYWY0YmYzOSoifV19&Signature=ZOf75uwDswP3jFkqCnXBCqs8ujWP5IM%7E8diNYD8PNwochvtwRt99DB5xEqfrqRuOzEC321IU5U1yKdUxUXRQtwSyVnZZwCdfE3b5jy0ZxTn2mFKtHm63auYyELXOz%7E8cm51Rdx3AQUdsXo-JaA1-oUFbYvqB6xUPW41nZX6JxbYkavxwDOVMb8PrajB5Zv5M3LyhCB7dyAzh3%7E%7E-Hv2Fdj7Gb9qlS4XIQcDlFBUQT9y33hYCZtXzKWWeLzOD8PjiB4OeG8GfxgcCfcC4Q5GbdXGOv67dTvYYCLfpQQ6PMaBVEqwchEcrEiinI6xInaAxhzUrgA6kHLWlV15fgDN68w__&Key-Pair-Id=K2L8F4GPSG1IFC)
- A journal entry in my own handwriting.

Both files are included in this folder so you can run the same tests.

## Plain OCR Testing
Testing plain OCR with no AI element to see what the baseline performance is.

### Easy OCR

In [ ]:
# Sample of test script
%pip install easyocr

import easyocr

reader = easyocr.Reader(["en"])

print("Reading receipt.png")
result = reader.readtext("receipt.png")
for bbox, text, prob in result:
    print(text, prob)

print("Reading journal.png")
result = reader.readtext("journal.png")
for bbox, text, prob in result:
    print(text, prob)

### Tesseract
Command line was used for these tests. I suggest you run them in your own command line, not in this workbook.

Tesseract first has to be installed: `brew install tesseract-lang`

It can then be used like so: `tesseract journal.png output.txt --tessdata-dir tessdata_best -l eng`

The results are output to `output.txt`.

### Thoughts
#### EasyOCR
- Easy to set up.
- Runs in python with minimal effort.

Receipt:
- Results seem mostly accurate, but no option for formatting.
- Some 0s missed.
```
[REG] BLACK SAKURA 0.7134233573496015
45,455 0.5145130493826988
COOKIE DOH SAUCES 0.9848700993099802
NATA DE COCO 0.9832491964194235
Sub Total 0.6755471839122579
45,455 0.8603865849225446
PB1 (16%) 0.60648566725297
545 0.9106112122535706
Rounding 0.9999930457599797
Total 0.9990530342745172
50,000 0.9971710126276468
Card Payment 0.9346906214744339
50,00q 0.46780132396508795
```

Journal:
- Not even worth recording. Didn't get any words correctly.

#### Tesseract
- Not good results with either receipt (blank) or journal entry (gibberish). 

## VLM Testing

### How VLM Works
1. First stream of input: The image
	1. Vision Encoder: Processes image into Feature Vectors, similar to an embedding. Picks up details like colours, edges, etc.
	2. Feature Vectors are fed to a Projector. This maps them to image tokens.
2. Second stream of input: A text prompt. This is eventually morphed into text tokens.
3. Image tokens and text tokens fed to LLM, which handles the answering of prompt.

Difficulties:
- Token Bottleneck: Complicated images might have details that are compressed to embedding size, losing detail.
- VLM can hallucinate.
- Bias from training data. Need to find the correct model trained on what we use it for.
- Content Security policies: VLMs could have security flags that would result in unprocessed data. Ideally we don't have these in place, but 3rd party services might have them included.

Difference from OCR:
- Visual Question Answering (VQA): You can ask it about text and images, not just a transcription.
- Can caption images based on content
- Document Understanding: Can summarize.

### Models Tested
The same two images were used. These were the prompts provided:

**Receipt**: *List the fields on this document and their values.*

**Journal Entry**: *Please transcribe all text in this image.*

| Model            | Parameters | Context Size | Model Size | Source      | Notes                                                                             |
| ---------------- | ---------- | ------------ | ---------- | ----------- | --------------------------------------------------------------------------------- |
| gemma3           | 4B         | 128K         | 3.3GB      | Ollama      |                                                                                   |
| gemma3           | 27B        | 128K         | 17GB       | Ollama      |                                                                                   |
| qwen3-vl         | 8B         | 256K         | 6.1GB      | Ollama      | A thinking model.                                                                 |
| llava            | 7B         | 32K          | 4.7GB      | Ollama      |                                                                                   |
| mistral-small3.2 | 24B        | 128K         | 15GB       | Ollama      |                                                                                   |
| moondream        | 1.8B       | 2K           | 1.7GB      | Ollama      | Very small. Could run anywhere.                                                   |
| idefics2         | 8B         | 8K           | ~15GB      | Huggingface |                                                                                   |
| PaddleOCR-VLM    | ??         | 16K          | ~8GB       | Huggingface | Could not get this to run on Mac or Windows, even with containerized environment. |

### Testing Method

For the Ollama-provided models, the prompts were run using the Ollama CLI. 

You must have Ollama CLI installed for this to work.

The syntax for this looks like:

`ollama run <model-name> <file path> <prompt>`

An example of a command used to read the journal entry would then be:

`ollama run gemma3:27b ./journal.png Please transcribe the text in this image.`

Note: Ollama also has a [JavaScript library](https://github.com/ollama/ollama-js).

**Warning!**

Each model you test with will download that model to your system if it is not already found. As noted above, some models can be large, so ensure you have the available storage capacity before running the command.

### Subjective Results
| Model            | Receipt                                                                                  | Journal (Handwriting)                                                                                                                                                   |
| ---------------- | ---------------------------------------------------------------------------------------- | ----------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| gemma3 (4B)      | Identified all fields and values.                                                        | Not good. Got a lot of the words, but also missed a lot and has things jumbled up.                                                                                      |
| gemma3 (27B)     | Perfect recognition. Even noted the initial position of items and their values together. | A lot better than 4B model. Two words that weren't right *Three* -> *Two* and *bullshit* -> *bullet hell*.                                                              |
| qwen3-vl        | Perfect recognition.                                                                     | Near perfect. Two mistakes: the word *criticals* and the name *Byleth*.                                                                                                 |
| llava            | Hallucinations. Didn't get all fields, made up numbers.                                  | Entirely hallucinated. I don't understand how it was this wrong. It knew it was a note, but then wrote its own note.                                                    |
| mistral-small3.2 | Better than expected. Identified all fields and values.                                  | Very slow to generate. Several minutes. Mostly perfect though. Got the two words that qwen3-v1 messed up on, but messed up *leaning -> learning* and *decided -> done*. |
| moondream        | Not good. Mostly incomplete with wrong numerical values.                                 | Complete failure. Described the page as an object but didn't transcribe the text at all.                                                                                |
| idefics2         | So slow that I cancelled it. Over 15 mins waited.                                        | Took a really long time. More errors than I would like, but the general structure of the sentences is there.                                                            |

### Example Output
These are example outputs from the `qwen3-vl` model:

#### Receipt
Here are the fields and their corresponding values from the document:  

    1. **1 [REG] BLACK SAKURA** → 45,455  
    2. **1 COOKIE DOH SAUCES** → 0  
    3. **1 NATA DE COCO** → 0  
    4. **Sub Total** → 45,455  
    5. **PB1 (16%)** → 4,545  
    6. **Rounding** → 0  
    7. **Total** → 50,000  
    8. **Card Payment** → 50,000

This can also be formatted as JSON by simply modifying the prompt. Some fine-tuning could still be done here, but the possibilities are here.
```json
{
  "1 [REG] BLACK SAKURA": 45455,
  "1 COOKIE DOH SAUCES": 0,
  "1 NATA DE COCO": 0,
  "Sub Total": 45455,
  "PB1 (16%)": 4545,
  "Rounding": 0,
  "Total": 50000,
  "Card Payment": 50000
}
```


#### Journal

Very clear output here. No additional instructions on formatting were given.

> Dec 13th - Sun.  
Had a thought that it might just be a low battery from continued testing. Maybe just enough to power the disk but not trigger the solenoid or relay? Finished the Three Houses DLC. It was really tough. Survived because of a few lucky 
mitigals and leaning on Bryglett. We beat Donkey Kong 2 as well. I cannot understand how those games were aimed at kids with that intense bullshit difficulty.  

## Testing on Expected Forms
Three sample forms have been filled out. Let's try to get the best results from them using `qwen3-vl`.

First, we will try to use the same approach as above, with a minimally-structured prompt.

`ollama run qwen3-vl ./form1.jpg Transcribe the values of this form in key:value JSON.`

```json
{
  "form_title": "Monthly Report",
  "submission_instruction": "Avoid delays, submit by the 5th",
  "organization": "Ministry of Social Development and Poverty Reduction, British Columbia",
  "privacy_notice": "The personal information requested on this form is collected by the Ministry of Social Development and Poverty Reduction pursuant to sections 25(g) of the Freedom of Information and Protection of Privacy Act for the 
purpose of administering the Employment and Assistance Act and Employment and Assistance for Persons with Disabilities Act. If you have any questions about the collection, use or disclosure of this information, please contact the Ministry 
of Social Development and Poverty Reduction at 1-866-866-0800.",
  "section1": {
    "are_you_still_in_need_of_assistance": "Yes",
    "has_family_unit_received_or_disposed_of_assets": "Yes",
    "any_changes_to_shelter_costs": "Yes",
    "any_changes_in_dependents_or_persons_living_in_home": "No",
    "any_employment_changes": {
      "Applicant": "Yes",
      "Spouse": "No"
    },
    "are_you_attending_or_enrolled_in_school_or_training": {
      "Applicant": "No",
      "Spouse": "Yes"
    },
    "are_you_looking_for_work": {
      "Applicant": "No",
      "Spouse": "Yes"
    },
    "have_you_moved_or_entered_facility": {
      "Applicant": "Yes",
      "Spouse": "No"
    },
    "any_outstanding_warrants_for_arrest": {
      "Applicant": "Yes",
      "Spouse": "No"
    },
    "changes_explanation": "Working part-time at Save-on-Foods, 20 hours a week. Started at 12. Pay stubs attached"
  },
  "section2": {
    "net_employment_income": {
      "Applicant": "$0",
      "Spouse": "$0"
    },
    "employment_insurance": {
      "Applicant": "$0",
      "Spouse": "$0"
    },
    "spousal_support_alimony": {
      "Applicant": "$0",
      "Spouse": "$0"
    },
    "child_support": {
      "Applicant": "$0",
      "Spouse": "$0"
    },
    "workbc_financial_support": {
      "Applicant": "$0",
      "Spouse": "$0"
    },
    "student_funding": {
      "Applicant": "$0",
      "Spouse": "$0"
    },
    "rental_income": {
      "Applicant": "$0",
      "Spouse": "$0"
    },
    "room_board_income": {
      "Applicant": "$0",
      "Spouse": "$0"
    },
    "workers_compensation": {
      "Applicant": "$0",
      "Spouse": "$0"
    },
    "private_pensions": {
      "Applicant": "$742.89",
      "Spouse": "$0"
    },
    "oas_gis": {
      "Applicant": "$0",
      "Spouse": "$0"
    },
    "canada_pension_plan_cpp": {
      "Applicant": "$682.39",
      "Spouse": "$0"
    },
    "tax_credits": {
      "Applicant": "$0",
      "Spouse": "$0"
    },
    "child_tax_benefits": {
      "Applicant": "$0",
      "Spouse": "$0"
    },
    "income_tax_refund": {
      "Applicant": "$0",
      "Spouse": "$0"
    },
    "all_other_income_money_received": {
      "Applicant": "$0",
      "Spouse": "$0"
    },
    "income_of_dependent_children": {
      "Applicant": "$0",
      "Spouse": "$0"
    }
  },
  "section3": {
    "applicant_signature": "Shaw",
    "applicant_print_name": "Sarah Thompson",
    "applicant_telephone": "234-567-890",
    "applicant_social_insurance_number": "234-456-112",
    "spouse_signature": "",
    "spouse_print_name": "",
    "spouse_telephone": "",
    "spouse_social_insurance_number": "",
    "declaration_date": "2025-11-11"
  },
  "sample_note": "Sample"
}
```

It made some mistakes, but it picked up most of the fields as expected. 

With some additional prompt engineering, it could be better.

Let's specify exactly what fields we want it to find and how their values should be formatted.

In [1]:
# Install ollama if not already installed
%pip install ollama

  Using cached httpx-0.28.1-py3-none-any.whl.metadata (7.1 kB)
  Using cached annotated_types-0.7.0-py3-none-any.whl.metadata (15 kB)
  Using cached sniffio-1.3.1-py3-none-any.whl.metadata (3.9 kB)
Using cached httpx-0.28.1-py3-none-any.whl (73 kB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.9/1.9 MB 4.4 MB/s eta 0:00:00a 0:00:01
Using cached annotated_types-0.7.0-py3-none-any.whl (13 kB)
Using cached sniffio-1.3.1-py3-none-any.whl (10 kB)

[notice] A new release of pip is available: 24.2 -> 25.3
[notice] To update, run: python3.11 -m pip install --upgrade pip
Note: you may need to restart the kernel to use updated packages.


In [15]:
def run_ollama(image_path, prompt, model):
    from ollama import generate

    response = generate(
        model=model,
        images=[image_path],
        prompt=prompt,
        stream=False
    )
    return response.response

In [ ]:
prompt = """
Transcribe the values of this form in key:value JSON.
Only return the JSON object.

Follow these instructions when extracting values:
- For date fields, use the format YYYY-MM-DD.
- For Yes/No checkboxes, use booleans. true if marked, false if not marked.
- The checkbox must be actually marked. A Yes checkbox with no mark is false. 
- For numeric fields, return only the number without any units or extra text.
- If a non-numeric value is present in a numeric field, use 0.
- If the number is missing, use 0.
- For telephone numbers, expect this format: xxx-xxx-xxxx
- Do not modify the field names or structure of the JSON object.
- Do not change the order of the fields in the JSON object.
- Do not enclose the JSON object in any additional text or code blocks.

The fields to extract and the structure of the JSON object are as follows:
{
  "section1": {
    "Are you still in need of assistance": boolean,
    "Has your family unit received or disposed of any assets?": boolean,
    "Any changes to your shelter costs?": boolean,
    "Any changes to Dependants or Persons living in the home?": boolean,
    "Any employment changes?": {
      "Applicant": boolean,
      "Spouse": boolean
    },
    "Are you attending / enrolled in school or training?": {
      "Applicant": boolean,
      "Spouse": boolean
    },
    "Are you looking for work?": {
      "Applicant": boolean,
      "Spouse": boolean
    },
    "Have you moved or entered a facility?": {
      "Applicant": boolean,
      "Spouse": boolean
    },
    "Any outstanding warrants for your arrest?": {
      "Applicant": boolean,
      "Spouse": boolean
    },
    "Please explain all changes including income and submit proof": string
  },
  "section2": {
    "Net Employment Income": {
      "Applicant": number,
      "Spouse": number
    },
    "Employmnet Insurance": {
      "Applicant": number,
      "Spouse": number
    },
    "Spousal Support / Alimony": {
      "Applicant": number,
      "Spouse": number
    },
    "Child Support": {
      "Applicant": number,
      "Spouse": number
    },
    "WorkBC Financial Support": {
      "Applicant": number,
      "Spouse": number
    },
    "Student Funding": {
      "Applicant": number,
      "Spouse": number
    },
    "Rental Income": {
      "Applicant": number,
      "Spouse": number
    },
    "Room / Board Income": {
      "Applicant": number,
      "Spouse": number
    },
    "Worker's Compensation": {
      "Applicant": number,
      "Spouse": number
    },
    "Private Pensions": {
      "Applicant": number,
      "Spouse": number
    },
    "OAS / GIS": {
      "Applicant": number,
      "Spouse": number
    },
    "Trust Income": {
      "Applicant": number,
      "Spouse": number
    },
    "Canada Pension Plan": {
      "Applicant": number,
      "Spouse": number
    },
    "Tax Credits": {
      "Applicant": number,
      "Spouse": number
    },
    "Child Tax Benefits": {
      "Applicant": number,
      "Spouse": number
    },
    "Income Tax Refund": {
      "Applicant": number,
      "Spouse": number
    },
    "All other income / money received": {
      "Applicant": number,
      "Spouse": number
    },
    "Income of Dependant Children": {
      "Applicant": number,
      "Spouse": number
    }
  },
  "section3": {
    "Applicant Signature": string,
    "Applicant Date": string,
    "Applicant Print Name": string,
    "Applicant Telephone": string,
    "Applicant Social Insurance Number": string,
    "Spouse Signature": string,
    "Spouse Date": string,
    "Spouse Print Name": string,
    "Spouse Telephone": string,
    "Spouse Social Insurance Number": string
  }
}
"""

In [77]:
def pretty_print(str_response):
    import json
    try:
        return json.loads(str_response.strip())
    except json.JSONDecodeError as e:
        # Try to remove first and last lines, as it might be in a code block
        lines = str_response.strip().split('\n')
        if len(lines) > 2:
            try:
                return json.loads('\n'.join(lines[1:-1]))
            except json.JSONDecodeError:
                pass
        print(str_response)
        raise e

In [29]:
response = run_ollama("./form1.jpg", prompt, "qwen3-vl")

In [30]:
pretty_print(response)

{'section1': {'Are you still in need of assistance': True,
  'Has your family unit received or disposed of any assets?': True,
  'Any changes to your shelter costs?': True,
  'Any changes to Dependants or Persons living in the home?': False,
  'Any employment changes?': {'Applicant': True, 'Spouse': False},
  'Are you attending / enrolled in school or training?': {'Applicant': False,
   'Spouse': True},
  'Are you looking for work?': {'Applicant': False, 'Spouse': False},
  'Have you moved or entered a facility?': {'Applicant': True,
   'Spouse': False},
  'Any outstanding warrants for your arrest?': {'Applicant': True,
   'Spouse': False},
  'Please explain all changes including income and submit proof': 'Working part-time at Save-on-Foods, 20 hours a week. Started at 12. Pay stubs attached'},
 'section2': {'Net Employment Income': {'Applicant': 0, 'Spouse': 0},
  'Employmnet Insurance': {'Applicant': 0, 'Spouse': 0},
  'Spousal Support / Alimony': {'Applicant': 0, 'Spouse': 0},
  'Ch

### Initial Impressions
The results aren't perfect. 

Even after a few attempts and tweaks, there's a few common errors:
- It really wants to mark the spouse checkboxes as true (Yes), even when nothing is marked on the page.
- It never gets `Started Oct 12.`, instead converting `Oct` to `at`.
- The signature is nowhere close, but that's sort of expected. Humans can barely read those too.
- It misses a number on the phone number. Something about those 8s throws it off.
- It messes up 4s and 9s in this person's writing.
- The monetary values in section 2 don't always end up in the right category.

Let's try a more objective approach.

### Objective Testing

Below is a sample of the ideal output. It has all the fields populated with the correct answers. We'll compare based on the following criteria:

- If the field is missing entirely or the key simply doesn't match, -3 points.
- If the key is there, but the value is missing, -2 points. The opposite also applies here.
- If the key and value are there, but the value isn't correct, -0.1 points for every incorrect character.
- For booleans, if the wrong boolean is present, -1 point.
- Numbers are treated as strings for the purpose of this scoring.
- Models that fail to return valid JSON fail the task with a -10 score.

For the purposes of comparing results, a number closer to zero (higher) is better.


In [36]:
# Form 1 Ideal Output
form_1_ideal_output = {
  "section1": {
    "Are you still in need of assistance": True,
    "Has your family unit received or disposed of any assets?": True,
    "Any changes to your shelter costs?": True,
    "Any changes to Dependants or Persons living in the home?": False,
    "Any employment changes?": {
      "Applicant": True,
      "Spouse": False,
    },
    "Are you attending / enrolled in school or training?": {
      "Applicant": False,
      "Spouse": False,
    },
    "Are you looking for work?": {
      "Applicant": False,
      "Spouse": False,
    },
    "Have you moved or entered a facility?": {
      "Applicant": True,
      "Spouse": False,
    },
    "Any outstanding warrants for your arrest?": {
      "Applicant": True,
      "Spouse": False,
    },
    "Please explain all changes including income and submit proof": "Working part-time at Save-on-Foods, 20 hours a week. Started Oct 12. Pay stubs attached",
  },
  "section2": {
    "Net Employment Income": {
      "Applicant": 0,
      "Spouse": 0,
    },
    "Employmnet Insurance": {
      "Applicant": 0,
      "Spouse": 0,
    },
    "Spousal Support / Alimony": {
      "Applicant": 0,
      "Spouse": 0,
    },
    "Child Support": {
      "Applicant": 0,
      "Spouse": 0,
    },
    "WorkBC Financial Support": {
      "Applicant": 0,
      "Spouse": 0,
    },
    "Student Funding": {
      "Applicant": 0,
      "Spouse": 0,
    },
    "Rental Income": {
      "Applicant": 0,
      "Spouse": 0,
    },
    "Room / Board Income": {
      "Applicant": 0,
      "Spouse": 0,
    },
    "Worker's Compensation": {
      "Applicant": 0,
      "Spouse": 0,
    },
    "Private Pensions": {
      "Applicant": 0,
      "Spouse": 0,
    },
    "OAS / GIS": {
      "Applicant": 742.84,
      "Spouse": 0,
    },
    "Trust Income": {
      "Applicant": 0,
      "Spouse": 0,
    },
    "Canada Pension Plan": {
      "Applicant": 682.34,
      "Spouse": 0,
    },
    "Tax Credits": {
      "Applicant": 0,
      "Spouse": 0,
    },
    "Child Tax Benefits": {
      "Applicant": 0,
      "Spouse": 0,
    },
    "Income Tax Refund": {
      "Applicant": 0,
      "Spouse": 0,
    },
    "All other income / money received": {
      "Applicant": 0,
      "Spouse": 0,
    },
    "Income of Dependant Children": {
      "Applicant": 0,
      "Spouse": 0,
    }
  },
  "section3": {
    "Applicant Signature": "SThompson",
    "Applicant Date": "2025-11-11",
    "Applicant Print Name": "Sarah Thompson",
    "Applicant Telephone": "234-567-8890",
    "Applicant Social Insurance Number": "234-456-112",
    "Spouse Signature": "",
    "Spouse Date": "",
    "Spouse Print Name": "",
    "Spouse Telephone": "",
    "Spouse Social Insurance Number": ""
  }
}

In [40]:
MISSING_KEY_PENALTY = -3
MISSING_VALUE_PENALTY = -2
BOOLEAN_MISMATCH_PENALTY = -1
CHARACTER_MISMATCH_PENALTY = -0.1

def compare_strings(str1, str2):
    # Ensure both are strings
    str1 = str(str1).strip()
    str2 = str(str2).strip()
    # If one string is blank and the other is not...
    if (not str1 and str2) or (str1 and not str2):
        return MISSING_VALUE_PENALTY  # One is missing
    # Compare two strings by how many characters differ
    if str1 == str2:
        return 0  # Exact match
    diff_count = sum(c1 != c2 for c1, c2 in zip(str1, str2))
    # Add difference for length mismatch
    diff_count += abs(len(str1) - len(str2))
    return CHARACTER_MISMATCH_PENALTY * diff_count  # Penalty per differing character

def compare_booleans(bool1, bool2):
    if bool1 == bool2:
        return 0  # Exact match
    else:
        return BOOLEAN_MISMATCH_PENALTY  # Mismatch
    
# Iterate through all keys in the ideal output and compare with the response
def calculate_penalties(response, ideal_response):
    total_score = 0
    for key, ideal_value in ideal_response.items():
        if key not in response:
            total_score += MISSING_KEY_PENALTY
            continue
        response_value = response[key]
        if isinstance(ideal_value, dict):
            # Recursive comparison for nested dictionaries
            total_score += calculate_penalties(response_value, ideal_value)
        elif isinstance(ideal_value, bool):
            total_score += compare_booleans(response_value, ideal_value)
        else:
            total_score += compare_strings(response_value, ideal_value)
    # Return score to nearest tenth
    return round(total_score, 1)

In [42]:
penalties = calculate_penalties(pretty_print(response), form_1_ideal_output)
print(f"Final Score: {penalties}")

Final Score: -6.0


Now, let's repeat this process across all three forms and a few of the previous models.

I've broken them up into separate blocks so we can just run them at our choosing.

The models tested below:
- qwen3-vl:8b
- gemma3:4b
- gemma3:27b
- mistral-small3.2:24b

NOTE: If you haven't pulled these models locally, the Python code below won't automatically pull them. It will just be unable to find them.

Pull them using Ollama with this command: `ollama pull <model-name>`

In [45]:
# Ideal outputs for forms 2 and 3
form_2_ideal_output = {
  "section1": {
    "Are you still in need of assistance": True,
    "Has your family unit received or disposed of any assets?": False,
    "Any changes to your shelter costs?": False,
    "Any changes to Dependants or Persons living in the home?": False,
    "Any employment changes?": {
      "Applicant": False,
      "Spouse": False,
    },
    "Are you attending / enrolled in school or training?": {
      "Applicant": False,
      "Spouse": False,
    },
    "Are you looking for work?": {
      "Applicant": True,
      "Spouse": False,
    },
    "Have you moved or entered a facility?": {
      "Applicant": False,
      "Spouse": False,
    },
    "Any outstanding warrants for your arrest?": {
      "Applicant": False,
      "Spouse": False,
    },
    "Please explain all changes including income and submit proof": "Began looking for work",
  },
  "section2": {
    "Net Employment Income": {
      "Applicant": 0,
      "Spouse": 0,
    },
    "Employmnet Insurance": {
      "Applicant": 0,
      "Spouse": 0,
    },
    "Spousal Support / Alimony": {
      "Applicant": 0,
      "Spouse": 0,
    },
    "Child Support": {
      "Applicant": 0,
      "Spouse": 0,
    },
    "WorkBC Financial Support": {
      "Applicant": 0,
      "Spouse": 0,
    },
    "Student Funding": {
      "Applicant": 0,
      "Spouse": 0,
    },
    "Rental Income": {
      "Applicant": 0,
      "Spouse": 0,
    },
    "Room / Board Income": {
      "Applicant": 0,
      "Spouse": 0,
    },
    "Worker's Compensation": {
      "Applicant": 0,
      "Spouse": 0,
    },
    "Private Pensions": {
      "Applicant": 0,
      "Spouse": 0,
    },
    "OAS / GIS": {
      "Applicant": 0,
      "Spouse": 0,
    },
    "Trust Income": {
      "Applicant": 0,
      "Spouse": 0,
    },
    "Canada Pension Plan": {
      "Applicant": 0,
      "Spouse": 0,
    },
    "Tax Credits": {
      "Applicant": 0,
      "Spouse": 0,
    },
    "Child Tax Benefits": {
      "Applicant": 0,
      "Spouse": 0,
    },
    "Income Tax Refund": {
      "Applicant": 0,
      "Spouse": 0,
    },
    "All other income / money received": {
      "Applicant": 0,
      "Spouse": 0,
    },
    "Income of Dependant Children": {
      "Applicant": 0,
      "Spouse": 0,
    }
  },
  "section3": {
    "Applicant Signature": "JSmith",
    "Applicant Date": "2025-11-12",
    "Applicant Print Name": "John Smith",
    "Applicant Telephone": "123-456-7789",
    "Applicant Social Insurance Number": "123-456-789",
    "Spouse Signature": "MSmith",
    "Spouse Date": "2025-11-12",
    "Spouse Print Name": "Maria Smith",
    "Spouse Telephone": "234-567-8890",
    "Spouse Social Insurance Number": "456-789-012"
  }
}

form_3_ideal_output = {
  "section1": {
    "Are you still in need of assistance": False,
    "Has your family unit received or disposed of any assets?": True,
    "Any changes to your shelter costs?": False,
    "Any changes to Dependants or Persons living in the home?": False,
    "Any employment changes?": {
      "Applicant": True,
      "Spouse": False,
    },
    "Are you attending / enrolled in school or training?": {
      "Applicant": False,
      "Spouse": False,
    },
    "Are you looking for work?": {
      "Applicant": True,
      "Spouse": False,
    },
    "Have you moved or entered a facility?": {
      "Applicant": False,
      "Spouse": False,
    },
    "Any outstanding warrants for your arrest?": {
      "Applicant": False,
      "Spouse": False,
    },
    "Please explain all changes including income and submit proof": "No changes. Continue to receive CPP, OAS, private pension",
  },
  "section2": {
    "Net Employment Income": {
      "Applicant": 2546,
      "Spouse": 0,
    },
    "Employmnet Insurance": {
      "Applicant": 230,
      "Spouse": 0,
    },
    "Spousal Support / Alimony": {
      "Applicant": 0,
      "Spouse": 0,
    },
    "Child Support": {
      "Applicant": 0,
      "Spouse": 0,
    },
    "WorkBC Financial Support": {
      "Applicant": 0,
      "Spouse": 0,
    },
    "Student Funding": {
      "Applicant": 0,
      "Spouse": 0,
    },
    "Rental Income": {
      "Applicant": 0,
      "Spouse": 0,
    },
    "Room / Board Income": {
      "Applicant": 0,
      "Spouse": 0,
    },
    "Worker's Compensation": {
      "Applicant": 0,
      "Spouse": 0,
    },
    "Private Pensions": {
      "Applicant": 0,
      "Spouse": 0,
    },
    "OAS / GIS": {
      "Applicant": 0,
      "Spouse": 0,
    },
    "Trust Income": {
      "Applicant": 0,
      "Spouse": 0,
    },
    "Canada Pension Plan": {
      "Applicant": 0,
      "Spouse": 0,
    },
    "Tax Credits": {
      "Applicant": 0,
      "Spouse": 0,
    },
    "Child Tax Benefits": {
      "Applicant": 0,
      "Spouse": 0,
    },
    "Income Tax Refund": {
      "Applicant": 0,
      "Spouse": 0,
    },
    "All other income / money received": {
      "Applicant": 0,
      "Spouse": 0,
    },
    "Income of Dependant Children": {
      "Applicant": 0,
      "Spouse": 0,
    }
  },
  "section3": {
    "Applicant Signature": "JLee",
    "Applicant Date": "2025-07-13",
    "Applicant Print Name": "Jennifer Lee",
    "Applicant Telephone": "234-556-7890",
    "Applicant Social Insurance Number": "235-678-012",
    "Spouse Signature": "",
    "Spouse Date": "",
    "Spouse Print Name": "",
    "Spouse Telephone": "",
    "Spouse Social Insurance Number": ""
  }
}

In [62]:
model_scores_map = {}  # To store model scores across forms

In [71]:
def success_or_fail(func):
    try:
        return func()
    except Exception as e:
        print(f"Error occurred: {e}")
        return -1000  # Large negative score for failure

def get_model_form_scores(model_name):
  model_specific_map = {}
  form_1_score = success_or_fail(lambda: calculate_penalties(
    pretty_print(run_ollama("./form1.jpg", prompt, model_name)), 
    form_1_ideal_output))
  print('Form 1 Complete')
  form_2_score = success_or_fail(lambda: calculate_penalties(
      pretty_print(run_ollama("./form2.jpg", prompt, model_name)),
      form_2_ideal_output))
  print('Form 2 Complete')
  form_3_score = success_or_fail(lambda: calculate_penalties(
      pretty_print(run_ollama("./form3.jpg", prompt, model_name)),
      form_3_ideal_output))
  print('Form 3 Complete')
  model_specific_map['form1'] = form_1_score
  model_specific_map['form2'] = form_2_score
  model_specific_map['form3'] = form_3_score
  return model_specific_map

In [66]:
# - qwen3-vl:8b
model_scores_map["qwen3-vl:8b"] = get_model_form_scores("qwen3-vl:8b")

Form 1 Complete
Form 2 Complete
Form 3 Complete


In [78]:
# - gemma3:4b
model_scores_map["gemma3:4b"] = get_model_form_scores("gemma3:4b")

Form 1 Complete
Form 2 Complete
Form 3 Complete


In [79]:
# - gemma3:27b
model_scores_map["gemma3:27b"] = get_model_form_scores("gemma3:27b")

Form 1 Complete
Form 2 Complete
Form 3 Complete


In [82]:
# - mistral-small3.2:24b
model_scores_map["mistral-small3.2:24b"] = get_model_form_scores("mistral-small3.2:24b")

```json
{
  "section1": {
    "Are you still in need of assistance": true,
    "Has your family unit received or disposed of any assets?": true,
    "Any changes to your shelter costs?": true,
    "Any changes to Dependants or Persons living in the home?": false,
    "Any employment changes?": {
      "Applicant": true,
      "Spouse": false,
    },
    "Are you attending / enrolled in school or training?": {
      "Applicant": false,
      "Spouse": false,
    },
    "Are you looking for work?": {
      "Applicant": false,
      "Spouse": false,
    },
    "Have you moved or entered a facility?": {
      "Applicant": true,
      "Spouse": false,
    },
    "Any outstanding warrants for your arrest?": {
      "Applicant": true,
      "Spouse": false,
    },
    "Please explain all changes including income and submit proof": "Working part-time at Save-on-Foods, 20 hours a week. Started Oct 12. Paysubs attached"
  },
  "section2": {
    "Net Employment Income": {
      "Applicant": 0,
  

In [85]:
# Mistral failed to output valid JSON, adding trailing commas etc. Fixing manually here for scoring purposes.

mistral_form_1_cleaned_output = "{\n   \"section1\":{\n      \"Are you still in need of assistance\":true,\n      \"Has your family unit received or disposed of any assets?\":true,\n      \"Any changes to your shelter costs?\":true,\n      \"Any changes to Dependants or Persons living in the home?\":false,\n      \"Any employment changes?\":{\n         \"Applicant\":true,\n         \"Spouse\":false\n      },\n      \"Are you attending / enrolled in school or training?\":{\n         \"Applicant\":false,\n         \"Spouse\":false\n      },\n      \"Are you looking for work?\":{\n         \"Applicant\":false,\n         \"Spouse\":false\n      },\n      \"Have you moved or entered a facility?\":{\n         \"Applicant\":true,\n         \"Spouse\":false\n      },\n      \"Any outstanding warrants for your arrest?\":{\n         \"Applicant\":true,\n         \"Spouse\":false\n      },\n      \"Please explain all changes including income and submit proof\":\"Working part-time at Save-on-Foods, 20 hours a week. Started Oct 12. Paysubs attached\"\n   },\n   \"section2\":{\n      \"Net Employment Income\":{\n         \"Applicant\":0,\n         \"Spouse\":0\n      },\n      \"Employmnet Insurance\":{\n         \"Applicant\":0,\n         \"Spouse\":0\n      },\n      \"Spousal Support / Alimony\":{\n         \"Applicant\":0,\n         \"Spouse\":0\n      },\n      \"Child Support\":{\n         \"Applicant\":0,\n         \"Spouse\":0\n      },\n      \"WorkBC Financial Support\":{\n         \"Applicant\":0,\n         \"Spouse\":0\n      },\n      \"Student Funding\":{\n         \"Applicant\":0,\n         \"Spouse\":0\n      },\n      \"Rental Income\":{\n         \"Applicant\":0,\n         \"Spouse\":0\n      },\n      \"Room / Board Income\":{\n         \"Applicant\":0,\n         \"Spouse\":0\n      },\n      \"Worker's Compensation\":{\n         \"Applicant\":0,\n         \"Spouse\":0\n      },\n      \"Private Pensions\":{\n         \"Applicant\":0,\n         \"Spouse\":0\n      },\n      \"OAS / GIS\":{\n         \"Applicant\":742.87,\n         \"Spouse\":0\n      },\n      \"Trust Income\":{\n         \"Applicant\":0,\n         \"Spouse\":0\n      },\n      \"Canada Pension Plan\":{\n         \"Applicant\":682.39,\n         \"Spouse\":0\n      },\n      \"Tax Credits\":{\n         \"Applicant\":0,\n         \"Spouse\":0\n      },\n      \"Child Tax Benefits\":{\n         \"Applicant\":0,\n         \"Spouse\":0\n      },\n      \"Income Tax Refund\":{\n         \"Applicant\":0,\n         \"Spouse\":0\n      },\n      \"All other income / money received\":{\n         \"Applicant\":0,\n         \"Spouse\":0\n      },\n      \"Income of Dependant Children\":{\n         \"Applicant\":0,\n         \"Spouse\":0\n      }\n   },\n   \"section3\":{\n      \"Applicant Signature\":\"Sarah Thompson\",\n      \"Applicant Date\":\"2023-11-11\",\n      \"Applicant Print Name\":\"Sarah Thompson\",\n      \"Applicant Telephone\":\"234-567-8840\",\n      \"Applicant Social Insurance Number\":\"230-456-112\",\n      \"Spouse Signature\":\"\",\n      \"Spouse Date\":\"\",\n      \"Spouse Print Name\":\"\",\n      \"Spouse Telephone\":\"\",\n      \"Spouse Social Insurance Number\":\"\"\n   }\n}"
mistral_form_2_cleaned_output = "{\n   \"section1\":{\n      \"Are you still in need of assistance\":true,\n      \"Has your family unit received or disposed of any assets?\":false,\n      \"Any changes to your shelter costs?\":false,\n      \"Any changes to Dependants or Persons living in the home?\":false,\n      \"Any employment changes?\":{\n         \"Applicant\":false,\n         \"Spouse\":false\n      },\n      \"Are you attending / enrolled in school or training?\":{\n         \"Applicant\":false,\n         \"Spouse\":false\n      },\n      \"Are you looking for work?\":{\n         \"Applicant\":true,\n         \"Spouse\":false\n      },\n      \"Have you moved or entered a facility?\":{\n         \"Applicant\":false,\n         \"Spouse\":false\n      },\n      \"Any outstanding warrants for your arrest?\":{\n         \"Applicant\":false,\n         \"Spouse\":false\n      },\n      \"Please explain all changes including income and submit proof\":\"Began looking for work\"\n   },\n   \"section2\":{\n      \"Net Employment Income\":{\n         \"Applicant\":0,\n         \"Spouse\":0\n      },\n      \"Employmnet Insurance\":{\n         \"Applicant\":0,\n         \"Spouse\":0\n      },\n      \"Spousal Support / Alimony\":{\n         \"Applicant\":0,\n         \"Spouse\":0\n      },\n      \"Child Support\":{\n         \"Applicant\":0,\n         \"Spouse\":0\n      },\n      \"WorkBC Financial Support\":{\n         \"Applicant\":0,\n         \"Spouse\":0\n      },\n      \"Student Funding\":{\n         \"Applicant\":0,\n         \"Spouse\":0\n      },\n      \"Rental Income\":{\n         \"Applicant\":0,\n         \"Spouse\":0\n      },\n      \"Room / Board Income\":{\n         \"Applicant\":0,\n         \"Spouse\":0\n      },\n      \"Worker's Compensation\":{\n         \"Applicant\":0,\n         \"Spouse\":0\n      },\n      \"Private Pensions\":{\n         \"Applicant\":0,\n         \"Spouse\":0\n      },\n      \"OAS / GIS\":{\n         \"Applicant\":0,\n         \"Spouse\":0\n      },\n      \"Trust Income\":{\n         \"Applicant\":0,\n         \"Spouse\":0\n      },\n      \"Canada Pension Plan\":{\n         \"Applicant\":0,\n         \"Spouse\":0\n      },\n      \"Tax Credits\":{\n         \"Applicant\":0,\n         \"Spouse\":0\n      },\n      \"Child Tax Benefits\":{\n         \"Applicant\":0,\n         \"Spouse\":0\n      },\n      \"Income Tax Refund\":{\n         \"Applicant\":0,\n         \"Spouse\":0\n      },\n      \"All other income / money received\":{\n         \"Applicant\":0,\n         \"Spouse\":0\n      },\n      \"Income of Dependant Children\":{\n         \"Applicant\":0,\n         \"Spouse\":0\n      }\n   },\n   \"section3\":{\n      \"Applicant Signature\":\"John Smith\",\n      \"Applicant Date\":\"2025-11-12\",\n      \"Applicant Print Name\":\"John Smith\",\n      \"Applicant Telephone\":\"123-456-7789\",\n      \"Applicant Social Insurance Number\":\"123-456-789\",\n      \"Spouse Signature\":\"Maria Smith\",\n      \"Spouse Date\":\"2025-11-12\",\n      \"Spouse Print Name\":\"Maria Smith\",\n      \"Spouse Telephone\":\"224-567-8890\",\n      \"Spouse Social Insurance Number\":\"456-789-012\"\n   }\n}"
mistral_form_3_cleaned_output = "{\n   \"section1\":{\n      \"Are you still in need of assistance\":false,\n      \"Has your family unit received or disposed of any assets?\":true,\n      \"Any changes to your shelter costs?\":false,\n      \"Any changes to Dependants or Persons living in the home?\":false,\n      \"Any employment changes?\":{\n         \"Applicant\":true,\n         \"Spouse\":false\n      },\n      \"Are you attending / enrolled in school or training?\":{\n         \"Applicant\":false,\n         \"Spouse\":false\n      },\n      \"Are you looking for work?\":{\n         \"Applicant\":true,\n         \"Spouse\":false\n      },\n      \"Have you moved or entered a facility?\":{\n         \"Applicant\":false,\n         \"Spouse\":false\n      },\n      \"Any outstanding warrants for your arrest?\":{\n         \"Applicant\":false,\n         \"Spouse\":false\n      },\n      \"Please explain all changes including income and submit proof\":\"No changes, Continue to receive CPP, OAS, Private pension\"\n   },\n   \"section2\":{\n      \"Net Employment Income\":{\n         \"Applicant\":2576,\n         \"Spouse\":0\n      },\n      \"Employmnet Insurance\":{\n         \"Applicant\":270,\n         \"Spouse\":0\n      },\n      \"Spousal Support / Alimony\":{\n         \"Applicant\":0,\n         \"Spouse\":0\n      },\n      \"Child Support\":{\n         \"Applicant\":0,\n         \"Spouse\":0\n      },\n      \"WorkBC Financial Support\":{\n         \"Applicant\":0,\n         \"Spouse\":0\n      },\n      \"Student Funding\":{\n         \"Applicant\":0,\n         \"Spouse\":0\n      },\n      \"Rental Income\":{\n         \"Applicant\":0,\n         \"Spouse\":0\n      },\n      \"Room / Board Income\":{\n         \"Applicant\":0,\n         \"Spouse\":0\n      },\n      \"Worker's Compensation\":{\n         \"Applicant\":0,\n         \"Spouse\":0\n      },\n      \"Private Pensions\":{\n         \"Applicant\":0,\n         \"Spouse\":0\n      },\n      \"OAS / GIS\":{\n         \"Applicant\":0,\n         \"Spouse\":0\n      },\n      \"Trust Income\":{\n         \"Applicant\":0,\n         \"Spouse\":0\n      },\n      \"Canada Pension Plan\":{\n         \"Applicant\":0,\n         \"Spouse\":0\n      },\n      \"Tax Credits\":{\n         \"Applicant\":0,\n         \"Spouse\":0\n      },\n      \"Child Tax Benefits\":{\n         \"Applicant\":0,\n         \"Spouse\":0\n      },\n      \"Income Tax Refund\":{\n         \"Applicant\":0,\n         \"Spouse\":0\n      },\n      \"All other income / money received\":{\n         \"Applicant\":0,\n         \"Spouse\":0\n      },\n      \"Income of Dependant Children\":{\n         \"Applicant\":0,\n         \"Spouse\":0\n      }\n   },\n   \"section3\":{\n      \"Applicant Signature\":\"Jolee\",\n      \"Applicant Date\":\"2025-07-13\",\n      \"Applicant Print Name\":\"Jennifer Lee\",\n      \"Applicant Telephone\":\"234-556-7890\",\n      \"Applicant Social Insurance Number\":\"235-678-012\",\n      \"Spouse Signature\":\"\",\n      \"Spouse Date\":\"\",\n      \"Spouse Print Name\":\"\",\n      \"Spouse Telephone\":\"\",\n      \"Spouse Social Insurance Number\":\"\"\n   }\n}"

model_scores_map["mistral-small3.2:24b"] = {
    'form1': calculate_penalties(
        pretty_print(mistral_form_1_cleaned_output),
        form_1_ideal_output),
    'form2': calculate_penalties(
        pretty_print(mistral_form_2_cleaned_output),
        form_2_ideal_output),
    'form3': calculate_penalties(
        pretty_print(mistral_form_3_cleaned_output),
        form_3_ideal_output)
}

In [ ]:
# Print the model scores map
print(model_scores_map)

{'qwen3-vl:8b': {'form1': -10.8, 'form2': -3.9, 'form3': -1.8}, 'gemma3:4b': {'form1': -29.8, 'form2': -14.0, 'form3': -24.4}, 'gemma3:27b': {'form1': -33.4, 'form2': -20.0, 'form3': -14.8}, 'mistral-small3.2:24b': {'form1': -3.3, 'form2': -1.9, 'form3': -0.7}}


### Results
Some models showed that they have difficulty just outputting the JSON without a code block around it. We've adjusted for that in the code above, so that particular failure isn't accounted for in these scores.

The `mistral` model still failed to make valid JSON after further tweaks to the prompt, but a look at the JSON it produced was really promising, despite the long processing times. I've included its scores with cleaned-up JSON, but that's the caveat.

The scores for each model can be seen below, but it's also important to know that this will definitely change with each run, as they are non-deterministic results.

| Model                  | Time   | Form 1       | Form 2 | Form 3 |
| ---------------------- | ------ | ------------ | ------ | ------ |
| `qwen3-vl:8b`          | 18m    | -10.8        | -3.9   | -1.8   |
| `gemma3:4b`            | 1m 40s | -29.8        | -14.0  | -24.4  |
| `gemma3:27b`           | 8m 30s | -33.4        | -20.0  | -14.8  |
| `mistral-small3.2:24b`* | 49m 21s | -3.3 | -1.9 | -0.7  |

### TLDR 

`qwen3-vl:8b` did a decent job. Maybe not good enough to be reliable without human intervention. 

`gemma3:4b` was much faster, as was its bigger version, but it contained more errors.

The `mistral` model is a unique candidate. It failed to output valid JSON, it took forever, but it made the fewest mistakes. 

Overall, it select the model that falls on the time-accuracy scale where you most need it.